falcon/cluster/similarity.py

```diff
@@ -0,0 +1,198 @@
+import collections
+
+import numba as nb
+import numpy as np
+import pandas as pd
+import scipy.optimize
+import scipy.sparse
+import spectrum_utils.spectrum as sus
+
+
+SpectrumTuple = collections.namedtuple(
+    "SpectrumTuple", ["precursor_mz", "precursor_charge", "mz", "intensity"]
+)
+
+#  return type
+SimilarityTuple = collections.namedtuple(
+    "SimilarityTuple",
+    [
+        "score",
+        "matched_intensity",
+        "max_contribution",
+        "n_greq_2p",  # signals contributing >= 2% score
+        "matches",  # number of matches
+        "matched_indices",
+        "matched_indices_other",
+    ],
+)
+
+
+def cosine(
+    spectrum1: pd.Series,
+    spectrum2: pd.Series,
+    fragment_mz_tolerance: float,
+    allow_shift: bool,
+) -> SimilarityTuple:
+    """
+    Compute the cosine similarity between the given spectra.
+
+    Parameters
+    ----------
+    spectrum1 : pd.Series
+        The first spectrum as a pandas Series.
+    spectrum2 : pd.Series
+        The second spectrum as a pandas Series.
+    fragment_mz_tolerance : float
+        The fragment m/z tolerance used to match peaks.
+    allow_shift : bool
+        Boolean flag indicating whether to allow peak shifts or not.
+
+    Returns
+    -------
+    SimilarityTuple
+        A tuple consisting of the cosine similarity between both spectra,
+        matched intensity, maximum contribution by a signal pair, matched
+        signals, and arrays of the matching peak indexes in the first and
+        second spectrum.
+    """
+    spec_tup1 = _df_row_to_spectrum_tuple(spectrum1)
```

Comment:
**question:** Can we convert directly to spectrum tuples instead of having to make the roundtrip via a Series?

Reply:
The `cosine` function isn't even used anymore. I directly call `_cosine_fast`